### PR TITLE
Add Show Entire device feature

### LINKED ISSUE #12: Show entire device view on click of expand button

### PR DESCRIPTION
**Pull request by O-Susanoo**
This one is a very tricky to review but the basic idea is to get the minimum enclosing square of the device and translate to this one , there is a small bug which is a border case , it's when we put component at both extremes , than the grid is off by one row, i will probably work on that case later if we end-up having time for it in this week

closes #12

### CODE REVIEW

File: src/components/Configurator.tsx [hunk 1]
```diff
@@ -356,7 +408,7 @@ class Configurator extends React.Component<any, ConfiguratorState> {
               if ((lastPosI !== posI || lastPosJ !== posJ) && (lastPosI >= 0 && lastPosJ >= 0)) {
                 device.dragLeave(device.gridState[lastPosI][lastPosJ].rect)
               }
-
+              // bug here
```
Comment (O-Susanoo) [inline src/components/Configurator.tsx]: there is a bug here , if user drag rapidly we should check if posI and posJ are not negative

File: src/components/Configurator.tsx [hunk 2]
```diff
@@ -162,6 +162,58 @@ class Configurator extends React.Component<any, ConfiguratorState> {
...
+        const scaleBy = maxWH / maxLength
+        const maxHeight = maxSize * scaleBy - this.svgHeight
+        const maxWidth = maxSize * scaleBy - this.svgWidth
+        const topPos = this.getIntervalLimit(minI * this.boxSize * maxWH / maxLength, 0, maxHeight)
+        const leftPos = this.getIntervalLimit(minJ * this.boxSize * maxWH / maxLength, 0, maxWidth)
+        this.svg.call(this.zoom).call(this.zoom.transform, d3.zoomIdentity.translate(-leftPos, -topPos).scale(scaleBy))
+      } else {
+        this.svg.call(this.zoom).call(this.zoom.transform, d3.zoomIdentity.translate(-minJ * this.boxSize, -minI * this.boxSize).scale(1))
+      }
+    }
```
Comment (DanielSeehausen) [inline src/components/Configurator.tsx]: this is getting pretty hairy to read - consider breaking the big statements into simpler named functions which describe the effect. The conditionals on line 200 and 201 are good candidates for this!

File: src/components/Configurator.tsx (hunk 1 above)
Comment (DanielSeehausen) [reply src/components/Configurator.tsx]: if you haven't would you able to make a ticket for it or fix in this PR? 

If you want to leave a comment in the code warning of the bug, it should be a little more descriptive for a future reader (imagine someone who has never seen this code base before coming upon a 'bug here' :D)

File: src/components/Configurator.tsx [hunk 3]
```diff
@@ -162,6 +162,58 @@ class Configurator extends React.Component<any, ConfiguratorState> {
     }
   }
 
+  getEnclosingSquare = () => {
+    let minI = -1
+    let maxI = -1
+    let minJ = -1
+    let maxJ = -1
+
+    for (let i = 0; i < this.gridState.length; i++) {
```
Comment (DanielSeehausen) [inline src/components/Configurator.tsx]: forEach takes a second param that is the current index!

`this.gridState.forEach((_, colIdx) => { ... })`

File: src/components/Configurator.tsx [hunk 4]
```diff
@@ -162,6 +162,58 @@ class Configurator extends React.Component<any, ConfiguratorState> {
...
+      for (let j = 0; j < this.gridState[i].length; j++) {
+        if (this.gridState[i][j].module) {
+          if (minI > i || minI === -1) minI = i
+          if (maxI < i || maxI === -1) maxI = i
+
+          if (minJ > j || minJ === -1) minJ = j
+          if (maxJ < j || maxJ === -1) maxJ = j
+
+        }
```
Comment (DanielSeehausen) [inline src/components/Configurator.tsx]: please name these variables (are they column index and row index?)

File: src/components/Configurator.tsx [hunk 5]
```diff
@@ -162,6 +162,58 @@ class Configurator extends React.Component<any, ConfiguratorState> {
...
+    if (minI !== -1 && this.svg && this.zoom && this.svgWidth && this.gridLayer && this.svgHeight) {
+      if ((maxI - minI) + (maxJ - minJ) !== 0) {
+        const maxLength = Math.max((maxI - minI + 2), (maxJ - minJ + 2)) * this.boxSize
+        const maxWH = Math.max(this.svgWidth, this.svgHeight)
+        const maxSize = this.gridN * this.boxSize
+        const scaleBy = maxWH / maxLength
+        const maxHeight = maxSize * scaleBy - this.svgHeight
+        const maxWidth = maxSize * scaleBy - this.svgWidth
+        const topPos = this.getIntervalLimit(minI * this.boxSize * maxWH / maxLength, 0, maxHeight)
+        const leftPos = this.getIntervalLimit(minJ * this.boxSize * maxWH / maxLength, 0, maxWidth)
```
Comment (DanielSeehausen) [inline src/components/Configurator.tsx]: consider naming repeat and complex math calls (this is a lot to put in as an argument)

File: src/components/Configurator.tsx [hunk 6]
```diff
@@ -162,6 +162,58 @@ class Configurator extends React.Component<any, ConfiguratorState> {
...
+    }
+  }
+
+  getIntervalLimit = (x: number, a: number, b: number) => {
+    if (x >= a && x <= b) {
+      return x
+    }
+
+    if (x < a) return a
+    return b
```
Comment (DanielSeehausen) [inline src/components/Configurator.tsx]: another way to express this:
`return conditional ? a : b`

File: src/components/Configurator.tsx [hunk 7]
```diff
@@ -162,6 +162,58 @@ class Configurator extends React.Component<any, ConfiguratorState> {
...
+
```
Comment (DanielSeehausen) [inline src/components/Configurator.tsx]: im a newb and don't know what these variables are :/
are x, a, b common variables to represent

File: src/components/Configurator.tsx [hunk 8]
```diff
@@ -162,6 +162,58 @@ class Configurator extends React.Component<any, ConfiguratorState> {
     }
   }
 
+  getEnclosingSquare = () => {
+    let minI = -1
+    let maxI = -1
+    let minJ = -1
+    let maxJ = -1
```
Comment (DanielSeehausen) [inline src/components/Configurator.tsx]: javascript has -Infinity to communicate your use case of -1 here!

File: src/components/DeviceControls.tsx [hunk 9]
```diff
@@ -6,15 +6,17 @@ type DeviceActions = {
...
 
-const DeviceControls: React.FC<DeviceActions> = ({ zoomIn, zoomOut, download }: DeviceActions) => (
+const DeviceControls: React.FC<DeviceActions> = ({ zoomIn, zoomOut, download, showEntireDevice }: DeviceActions) => (
   <Container textAlign='center'>
     <div style={{ margin: '3.5rem' }}>
       <ModuleMenu
         zoomIn={zoomIn}
         zoomOut={zoomOut}
         download={download}
+        showEntireDevice={showEntireDevice}
```
Comment (ConorTighe1995) [inline src/components/DeviceControls.tsx]: Nice!

File: src/components/Configurator.tsx (hunk 1 above)
Comment (O-Susanoo) [reply src/components/Configurator.tsx]: It was quick fix so i just fixed it in this pr

File: src/components/Configurator.tsx (hunk 6 above)
Comment (O-Susanoo) [reply src/components/Configurator.tsx]: Laser eye

File: src/components/Configurator.tsx (hunk 8 above)
Comment (O-Susanoo) [reply src/components/Configurator.tsx]: yes they are row and col indexes , and i don't think that infinity values could be much different , i will stuck with -1 , but i can see the source of the confusion :+1:

File: src/components/Configurator.tsx (hunk 3 above)
Comment (O-Susanoo) [reply src/components/Configurator.tsx]: Fixed !!

File: src/components/Configurator.tsx (hunk 4 above)
Comment (O-Susanoo) [reply src/components/Configurator.tsx]: Done

File: src/components/Configurator.tsx (hunk 7 above)
Comment (O-Susanoo) [reply src/components/Configurator.tsx]: done

File: src/components/Configurator.tsx (hunk 5 above)
Comment (O-Susanoo) [reply src/components/Configurator.tsx]: done

File: src/components/Configurator.tsx [hunk 10]
```diff
@@ -190,6 +190,64 @@ class Configurator extends React.Component<any, ConfiguratorState> {
     }
   }
 
+  getEnclosingSquare = () => {
+    let minRow = Infinity
```
Comment (ConorTighe1995) [inline src/components/Configurator.tsx]: Would Number.POSITIVE_INFINITY or Number.NEGATIVE_INFINITY be better here?

**Comment by O-Susanoo** [comment]
Hey Daniel , thank you for taking the time to review the pr , your comments were delightful and interesting , i made small changes to the code but not everything and the reason why is because as i told you on slack there is a big pr coming that will clean this stuff up and where i will defiantly try to give a user friendly code , concerning the ifs most of them are there because typescript can sometimes be a pain so if it doesn't see that you are  checking a nullable value in the if block then it will still complain about using directly a nullable value for example suppose i do this

let a:number|null ;

function X(){
logic to access a will complain because a can become nullable
however if i do this
if(a){
// good
}
if i try to abstract that check in another function then it will also complain , 
}

so that's why i hope i can solve this stuff in the refactor pr where i will actually have a module just for the configurator logic itself 

Again thank you for taking the time , and totally agree with you , you need to survive one more nasty pr that will come after this one and hopefully the refactor pr will have your wishes come to life

**Pull request merged**